### PR TITLE
Gen ume

### PR DESCRIPTION
**Pull request by Sidney-Lisanza**
## Description
Brief description of changes made

## Type of Change
- [ ] Bug fix
- [ X] New feature  
- [ ] Documentation update
- [ ] Performance improvement
- [X ] Code refactoring

### CODE REVIEW

File: src/lobster/callbacks/_structure_decode.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
+import lightning
+import os
+import torch
+from lobster.model.latent_generator.io import writepdb
+from loguru import logger
+
+
+class StructureDecode(lightning.Callback):
```
Comment (karinazad) [inline src/lobster/callbacks/_structure_decode.py:8]: nit: could you name name `StructureDecodeCallback` so it's a bit more clear (same for the other callback)  and coluld you pls add docstrings explaining what this callback does + how to use

File: src/lobster/callbacks/_structure_decode.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    def __init__(self, structure_path: str = None, save_every_n: int = 1000):
+        self.STRUCTURE_PATH = structure_path
```
Comment (karinazad) [inline src/lobster/callbacks/_structure_decode.py:10]: why uppercase?

File: src/lobster/callbacks/_structure_decode.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        self.save_every_n = save_every_n
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        if not os.path.exists(f"{self.STRUCTURE_PATH}/decode"):
+            os.makedirs(f"{self.STRUCTURE_PATH}/decode")
```
Comment (karinazad) [inline src/lobster/callbacks/_structure_decode.py:14]: let's use `os.path` or `Path` since structure path can have an extra `/`

File: src/lobster/datasets/_structure_dataset.py [hunk 4]
```diff
@@ -136,7 +136,7 @@ def __init__(
         import lobster
 
         lobster.ensure_package("torch_geometric", group="lg-gpu (or --extra lg-cpu)")
-        lobster.ensure_package("icream", group="lg-gpu (or --extra lg-cpu)")
+        # lobster.ensure_package("icream", group="lg-gpu (or --extra lg-cpu)")
```
Comment (karinazad) [inline src/lobster/datasets/_structure_dataset.py:139]: was this meant to be commented out? if not needed, we can remove

File: src/lobster/model/latent_generator/cmdline/inference.py [hunk 5]
```diff
@@ -400,6 +400,12 @@ def instantiate_dict_cfg(cfg: DictConfig | None, verbose=False):
 
 
 def load_config(config_path: str, config_name: str, overrides: list[str] | None = None) -> DictConfig:
+    from hydra.core.global_hydra import GlobalHydra
```
Comment (karinazad) [inline src/lobster/model/latent_generator/cmdline/inference.py:403]: this should be top-level import

File: src/lobster/transforms/_structure_transforms.py [hunk 6]
```diff
@@ -319,9 +319,9 @@ def __call__(self, x: dict) -> dict:
 
 class StructureLigandTransform(BaseTransform):
     def __init__(self, max_length=512, rand_permute_ligand=False, **kwargs):
-        import lobster
+        # import lobster
 
-        lobster.ensure_package("torch_geometric", group="lg-gpu (or --extra lg-cpu)")
+        # lobster.ensure_package("torch_geometric", group="lg-gpu (or --extra lg-cpu)")
```
Comment (karinazad) [inline src/lobster/transforms/_structure_transforms.py:324]: ^

File: src/lobster/model/gen_ume/_gen_ume_sequence_structure_encoder_lightning_module.py [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+            if just_loss:
+                return loss
+            # apply loss weighting from weight_dict in loss_factory; setting to 0 b/c will need different way to set weights for different losses
+            # total_loss += self.loss_factory.weight_dict[loss2apply_] * loss
+            total_loss += 0 * loss
+            loss_dict[f"{split}_{loss2apply_}"] = loss
+
+        return total_loss, loss_dict
+
+    def compute_mlm_loss(self, batch: dict[str, Tensor]) -> Tensor:
```
Comment (karinazad) [inline src/lobster/model/gen_ume/_gen_ume_sequence_structure_encoder_lightning_module.py:243]: do we even need to have this function here?

File: src/lobster/model/gen_ume/_gen_ume_sequence_structure_encoder_lightning_module.py [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+    def step(
+        self, batch: dict[str, Tensor], batch_idx: int, split: Literal["train", "val"] = "train"
+    ) -> dict[str, Tensor]:
+        """Single training/val/test step of the model."""
+        # set losses
+        total_loss = 0.0
+        loss_dict = {}
+
+        # prep the input
+        with torch.no_grad():
```
Comment (karinazad) [inline src/lobster/model/gen_ume/_gen_ume_sequence_structure_encoder_lightning_module.py:345]: just curious, why would this produce gradients?

Comment (Sidney-Lisanza) [reply src/lobster/model/gen_ume/_gen_ume_sequence_structure_encoder_lightning_module.py:357]: when we encoding the structure to get the structure tokens, those params could influence the gradients

**Review by karinazad — APPROVED** [review]
